### PR TITLE
Add CancellationToken-accepting overloads for SendPingAsync

### LINKED ISSUE #67260: [API Proposal]: Add `Ping.SendPingAsync` overloads with `CancellationToken`
### Background and motivation

This was already approved in https://github.com/dotnet/runtime/issues/14336, but reverted in https://github.com/dotnet/runtime/pull/67253 because this requires some more work.

See https://github.com/dotnet/runtime/pull/64860#discussion_r836496120

### API Proposal

```cs
namespace System.Net.NetworkInformation {
    public class Ping {
        public Task<PingReply> SendPingAsync(IPAddress address, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
        public Task<PingReply> SendPingAsync(string hostNameOrAddress, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
    }
}
```

/cc @danmoseley 
/cc @stephentoub

### PR DESCRIPTION
**Pull request by madelson**
Provides "true" cancellation for async ping methods using either a token or
the existing SendAsyncCancel() API.

Fix #67260

### CODE REVIEW

File: src/libraries/System.Net.Ping/ref/System.Net.Ping.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: As mentioned [here](https://github.com/dotnet/runtime/issues/67260#issuecomment-1185972901), I think this API would be more useful if `timeout` were an optional parameter. Requiring it here when other overloads allow you to skip it will likely lead to either poorly-chosen timeouts or copying the default value.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs:85]: I figured it didn't hurt to also allow timing out / canceling while reading the output, but  calling it out in case it isn't desirable.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:24]: I'm surprised that this doesn't need to be `volatile` since it gets written to and read from on different threads, with the reads eschewing a lock. With the updated code the read pattern is unchanged (no lock), we now write to this outside of a lock sometimes (from the cancellation token registration); does that change anything?

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:446]: Previously, cancellations using `SendAsyncCancel()` would get wrapped in `PingException` in the try-catch below. That didn't seem very desirable to me, so I changed it to return a canceled task regardless of how cancellation occurred. Curious what others think.

Other options would be to always wrap, or preserve the wrapping behavior for `SendAsyncCancel()` only.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:459]: TODO this should be more defensive:

```
when (e is not PlatformNotSupportedException && !(e is OperationCanceledException && _canceled))
```

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:440]: I did this to reduce logic duplication between the `IPAddress` and `string` paths, but let me know if I should unwind it to keep things how they were before.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs
Comment (rzikm) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs:88]: We should still discriminate by the exception type. If e.g. `ParsePingUtilityOutput` throws, we may end up returning PingReply with timeout.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (rzikm) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:446]: That would probably be a breaking change in the behavior, but since we are implementing "true" cancellation and change the behavior, it might be okay (and we just need to make sure it is documented).

Tagging @stephentoub for more thoughts.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (rzikm) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:24]: Since this is always set together with cancelling the _`timeoutOrCancellationSource`, I think we can get away without volatile, but I will defer to the opinion of the more experienced. cc: @stephentoub

**Review by rzikm — COMMENTED** [review]
Generally looks good, besides a few questions.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:446]: > we just need to make sure it is documented

@rzikm Do you mean in the doc comments of the new APIs?

> since we are implementing "true" cancellation and change the behavior, it might be okay

FWIW as a consumer of .NET async APIs, it is frustrating when cancellations get wrapped to look like some other type of exception.

File: src/libraries/System.Net.Ping/ref/System.Net.Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: @rzikm thoughts on this change?

**Comment by madelson** [comment]
@rzikm I thought I had asked about this in another comment, but I can't find it now. One callout is that this **does not implement true cancellation on Windows** because Windows pings [use a native function](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs#L170) which wasn't obviously cancelable from my read of the docs (let me know if I missed something). Maybe this is something that can be handled by [CancelIOEx](https://docs.microsoft.com/en-us/windows/win32/fileio/cancelioex-func)?

One other option I considered was changing the Windows ping to use raw sockets when possible (as is [done for Unix](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L32)). However, I'm not sure of the pros/cons there.

Thoughts?

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (rzikm) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:446]: > Do you mean in the doc comments of the new APIs?

I meant in release notes that are compiled and list breaking changes. It would be handled by us. AFAIK, we don't list `TaskCancelledException` in docs for async methods.

File: src/libraries/System.Net.Ping/ref/System.Net.Ping.cs
Comment (rzikm) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: I would stick to what was approved during the API review, any changes to that need to be green-lighted by the API review group.

**Comment by rzikm** [comment]
> One callout is that this does not implement true cancellation on Windows because Windows pings [use a native function](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs#L170) which wasn't obviously cancelable from my read of the docs (let me know if I missed something). Maybe this is something that can be handled by [CancelIOEx](https://docs.microsoft.com/en-us/windows/win32/fileio/cancelioex-func)?

It would be nice to have consistent behavior between platforms if it can be achieved, I did a quick test and CancelIO didn't work (returned INVALID_HANDLE). Not sure if using raw ICMP sockets is viable, maybe @antonfirsov will know.

File: src/libraries/System.Net.Ping/ref/System.Net.Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: @rzikm I'm fine to keep going with the approved API if that's the direction you want to take.  Just out of curiosity 

* What is the process for getting such a change re-reviewed?
* API review group aside, what is your personal view on which version is better? In my view switching to `TimeSpan? timeout = null` is quite a large improvement because as a caller I wouldn't know what a "reasonable" default timeout was. Interested to hear other opinions, though.

Comment (stephentoub) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: _Many_ APIs throughout the core libraries accept a timeout, none to my knowledge accept a nullable TimeSpan, and we wouldn't diverge for this one method. There's not really a reasonable default this method could pick other than infinite or system-default.

Comment (stephentoub) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: > as a caller I wouldn't know what a "reasonable" default timeout was

A user has a much better chance of picking an appropriate timeout for their use case than we do.

Comment (madelson) [reply src/libraries/System.Net.Ping/ref/System.Net.Ping.cs:65]: Makes sense thanks @stephentoub , especially if this would be a new pattern. I'll note that other existing Ping overloads *do* use a default timeout (5s), but it sounds like we're moving away from that rather than trying to match it.

**Comment by stephentoub** [comment]
> One callout is that this does not implement true cancellation on Windows because Windows pings [use a native function](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs#L170) which wasn't obviously cancelable from my read of the docs

I've not read the change yet. What do you do in this case? If it's just not as cancelable, e.g. once the native call is made the cancellation token is ignored, that's ok. If instead cancellation triggers the returned task to complete even while the native operation continues, that's not ok.

**Comment by stephentoub** [comment]
> One other option I considered was changing the Windows ping to use raw sockets when possible (as is [done for Unix](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L32)). However, I'm not sure of the pros/cons there.

IIRC, creating raw sockets requires elevated/admin privileges. If that's the case, we can't do that.

**Comment by madelson** [comment]
@stephentoub 

> I've not read the change yet. What do you do in this case? If it's just not as cancelable, e.g. once the native call is made the cancellation token is ignored, that's ok. If instead cancellation triggers the returned task to complete even while the native operation continues, that's not ok.

It is much closer to the former than the latter but not exactly. Before this change, we already had a (mostly non-functional) version of cancellation via the `Ping.SendAsyncCancel()` method which would just set the `_canceled` flag. When the async operation completed, [we'd convert to a canceled result based on the task](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs#L290). **I didn't change this code at all for the Windows branch.** The only change in the behavior is that previously cancellations were getting wrapped as `PingException` and now cancellations flow through without wrapping so that all cancellations manifest in the same way.

> IIRC, creating raw sockets requires elevated/admin privileges. If that's the case, we can't do that.

In Unix, we [use the raw socket version conditionally based on whether we have permissions](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L32). I was proposing that we do the same thing for Windows. Maybe you're just saying that people running in admin mode is so rare that having this fork isn't worth it?

**Review by rzikm — APPROVED** [review]
LGTM now, but I would like @stephentoub to have a look as well in case I missed something.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs:81]: var => bool

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs:91]: We don't want to only catch OperationCanceledException?  If the wait failed with an unrelated exception at appx the same time that cancellation was requested, we'll end up eating that exception and treating it as a timeout?

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:152]: Dispose doesn't Cancel.  Do you want to also Cancel?  I'm just wondering about the comment on this method which suggests it does so.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:484]: It's unlikely to make a material difference, but it looks like this could use UnsafeRegister instead of Register.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs:281]: Move this down into the try block?

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:88]: You should be able to mark this method as `[MemberNotNull(nameof(_timeoutOrCancellationSource))]`, which should let you remove at least one if not more `!`s on use of that field.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:489]: Doing this here means the timeout won't apply to the GetHostAddressesAsync inside getAddress.  Why is that ok / desirable?  I'm also unclear as to why this is being done like this after the SendPingAsyncCore.  Normally this kind of ordering would be to avoid a race condition, but that doesn't apply here.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:88]: Do you mean other methods? This is the one that lazily initializes the `CancellationTokenSource` (`Finish` then nulls it out if it has become canceled).

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs
Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.PingUtility.cs:91]: > Were you imagining

The latter.

> In that case, does it matter? We have a race condition where multiple things happened that should have aborted the ping and we are correctly informing the caller about one of those things.

When this situation typically arises elsewhere,  we make the same argument,  but in those cases there's the added complication that the cancellation could have actually caused the other exception,  and then it's actually preferable to show the cancelation one because the other is fake in a sense.  Here,  though, I don't think that's the case, in which case we generally want to prioritize real failures over manufactured ones from timeouts and cancellation,  as they typically convey better information about the state of the system.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:88]: I mean putting that attribute on this method should let you remove the `!` from other methods that call this one.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:152]: Interesting. I feel like this comment is just wrong. The way the method used to and still works is that if called while a ping request is in progress it sets `_disposeRequested` to true and then returns. Then when the ping request completes it calls `Finish` which completes the disposal.

It would be interesting if this comment implied that what `InternalDisposeCore` does (disposing the handles) can actually safely cancel outstanding Windows ping requests, but given that the [old .NET Framework version](https://referencesource.microsoft.com/#System/net/System/Net/NetworkInformation/ping.cs,131) seems to have the same behavior I'm not convinced. 

Do you agree? Should I just remove the comment? Should we change the code to start canceling the token once we've requested disposal to hurry outstanding requests along?

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (madelson) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:489]: Both of these are attempts to maintain the existing behavior / be true to the documentation:

* The old code does not apply the timeout to `Dns.GetHostAddressesAsync` (see https://source.dot.net/#System.Net.Ping/System/Net/NetworkInformation/Ping.cs,646).
* The documentation for [the `timeout` parameter](https://docs.microsoft.com/en-us/dotnet/api/system.net.networkinformation.ping.sendpingasync?view=net-6.0#system-net-networkinformation-ping-sendpingasync(system-string-system-int32-system-byte()-system-net-networkinformation-pingoptions)) says: "The maximum number of milliseconds (*after sending the echo message*) to wait for the ICMP echo reply message". 

I feel like the timeout is supposed to be inherent to the ping request, rather than being a general timeout for the entire operation. In contrast, I think the `CancellationToken` is the way for the caller to abort the operation whenever they want. Thoughts?

**Comment by madelson** [comment]
Thanks for all the feedback @rzikm @stephentoub . At this point I'm just waiting on resolutions to a few open threads:

* https://github.com/dotnet/runtime/pull/72338#discussion_r922843165
* https://github.com/dotnet/runtime/pull/72338#discussion_r946258155
* https://github.com/dotnet/runtime/pull/72338#discussion_r946262570

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:25]: It doesn't. There's always a happens-before relationship involved, e.g. the flag is set before the asynchronous operation is started, or it's read only after the things that could have set it have demonstrably quiesced.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:154]: > I feel like this comment is just wrong.

Looks like you're right. You can just delete it.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs
Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:491]: I'll defer to @dotnet/ncl on the desired behavior here.  I expect it was documented that way because there was literally no way before for the Dns operations to be timed out.  But there is now.  The question then is whether we want to update the timeout to apply to the whole operation or just to the waiting for the ICMP echo reply message; I'm ok either way.

I've not looked; are we consistent with the timeout only applying to the ICMP echo reply across all implementations?

Comment (rzikm) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs:491]: Personally, I feel like it is better to keep the timeout only for the ICMP echo reply. As noted, users can use the `CancellationToken` to cancel the operation *x* ms from the start of it, and if we make the `timeout` parameter cover the DNS resolution as well, users may get different results depending on whether DNS was fast enough or was/wasn't cached.

**Comment by rzikm** [comment]
Looks like there is just one thing to address https://github.com/dotnet/runtime/pull/72338#discussion_r946258155 and we will be good to go, right? @madelson?

**Review by rzikm — APPROVED** [review]
LGTM

**Comment by rzikm** [comment]
Thanks for the contribution!

**Pull request merged**
